junit-jupiter-api/src/main/java/org/junit/jupiter/api/Assertions.java

```diff
@@ -1479,12 +1560,102 @@ private static void failArraysNotEqual(Object expected, Object actual, Deque<Int
 		return result;
 	}
 
-	private static String formatIndexes(Deque<Integer> indexes) {
-		if (indexes == null || indexes.isEmpty()) {
-			return "";
+	// --- assertIterableEquals helpers ----------------------------------
+
+	private static void assertIterableEquals(Iterable<?> expected, Iterable<?> actual, Deque<Integer> indexes,
+			Supplier<String> messageSupplier) {
+
+		if (expected == actual) {
+			return;
 		}
-		String indexesString = indexes.stream().map(Object::toString).collect(joining("][", "[", "]"));
-		return " at index " + indexesString;
+		assertIterablesNotNull(expected, actual, indexes, messageSupplier);
+
+		Iterator<?> expectedIterator = expected.iterator();
+		Iterator<?> actualIterator = actual.iterator();
+
+		int processed;
+		for (processed = 0; expectedIterator.hasNext() && actualIterator.hasNext(); processed++) {
+			Object expectedElement = expectedIterator.next();
+			Object actualElement = actualIterator.next();
+
+			if (expectedElement == actualElement) {
+				continue;
+			}
+
+			indexes.addLast(processed);
+			assertIterableElementsEqual(expectedElement, actualElement, indexes, messageSupplier);
+			indexes.removeLast();
+		}
+
+		assertIteratorsAreEmpty(expectedIterator, actualIterator, processed, indexes, messageSupplier);
+	}
+
+	private static void assertIterableElementsEqual(Object expected, Object actual, Deque<Integer> indexes,
+			Supplier<String> messageSupplier) {
+		if (expected instanceof Iterable && actual instanceof Iterable) {
+			assertIterableEquals((Iterable<?>) expected, (Iterable<?>) actual, indexes, messageSupplier);
+		}
+		else if (!Objects.equals(expected, actual)) {
+			if (expected == null && actual instanceof Iterable) {
```

Comment:
**Potential code duplication:** I wonder if the instanceof check for `Iterable` in the following two if statement is required? I seems, that it doesnt matter whether or not the underlying object is an `Iterable` at this point. Moreover, this would mean that we do have a code duplication of the method `assertIterablesNotNull`. Still, I wonder if that is the case. Can you tell, what the instanceof check is required for?